pkg/controller/machine/machine_controller.go

```diff
@@ -278,6 +286,34 @@ func (r *ReconcileMachine) isDeleteAllowed(machine *clusterv1.Machine) bool {
 	return node.UID != machine.Status.NodeRef.UID
 }
 
+// isDeleteNodeAllowed returns true only if the Machine's NodeRef is not nil
+// and if the Machine is not the last control plane node in the cluster.
+func (r *ReconcileMachine) isDeleteNodeAllowed(ctx context.Context, machine *clusterv1.Machine) (bool, error) {
+
+	// Cannot delete something that doesn't exist.
+	if machine.Status.NodeRef == nil {
+		return false, nil
+	}
+
+	// Get all of the machines that belong to this cluster.
+	machines, err := r.getMachines(ctx, machine)
+	if err != nil {
+		return false, err
+	}
+
+	controlPlaneMachines := util.GetControlPlaneMachines(machines)
+
+	// If this machine is a control plane node then only delete the NodeRef
+	// if there is another control plane node (besides this machine) left.
+	if util.IsControlPlaneMachine(machine) {
+		return len(controlPlaneMachines) > 1, nil
+	}
+
+	// Otherwise it's okay to delete the NodeRef if there's a single control
+	// plane node left.
+	return len(controlPlaneMachines) > 0, nil
```

Comment:
Shouldn't this return false?

Reply:
If there are no controlPlane machines (externally managed control plane), then this would return false even when it isn't an issue.

Reply:
Scenario:

- The machine we're deleting is not a control plane machine
- If there is at least 1 control plane machine, it's ok to delete the node for this non-control plane machine

Right?

Reply:
Makes sense, I misread the if condition above

Reply:
That was my thinking. At first I assumed that the last machine would always be a control plane machine, but then I realized it may not be. So that's why the number of required control plane machines differs based on whether the provided machine is a control plane machine.